### DISCUSSION TITLE
It is not possible to change the game script in an existing scenario / save

### DISCUSSION THREAD

**Discussion by LC-Zorg** [OpenTTD/OpenTTD#11709] [Features]
### Version of OpenTTD

12.2 (old bug)

### Expected result

There is a simple and effective way to remove a script from the game and replace it with another or leave the save / scenario without the script. The game works fine after the change.

### Actual result

There is currently no way, or at least not known to me (and most players), to remove or replace a game script.
For me, this is a very topical problem and it is one example where it is a major disadvantage. I was modifying one of the scenarios, spending a few hours (luckily only a few) on the changes, and then noticed that I chose the wrong version to rework, with the wrong script. In this case, all work goes to the bin and I have to start all over again. In another case, it turned out that the selected game script was defective - there was nothing I could do here either. There may be a method, but if so, it's hidden. Yet another very common example is when a player wants to continue offline a game that he started on a server where some script was used. Usually in such cases the script stops working completely, and further development of cities or stopping development is impossible.

### Steps to reproduce

Save the game on the ttdistas or citymania server and try after loading the game to continue it by developing selected cities.

**Comment by James103** [comment]
As a workaround, you can try renaming the file extension of the savegame so it reads `.scn` instead of `.sav`. Then, open up the game in the Scenario Editor to change the assigned Game Script. You may need to `set ai_developer_tools true` first.
Once you have made your changes, save the scenario and play it to continue your game.

**Comment by LC-Zorg** [comment]
I know this method, but the problem here is that replacing the script in the editor often causes errors later in the game...
![Replacing GS](https://user-images.githubusercontent.com/67662046/197871934-a42fdfbe-3ecb-4f92-ae93-c1af30e87b24.png)
while trying to delete the script will closing the game by fatal application failure.

**Comment by James103** [comment]
> while trying to delete the script will closing the game by fatal application failure.

Do you have the crash files (at least `crash.log`, ideally also `crash.dmp`, `crash.sav`, and/or `crash.png`)? If so, can you please zip the files and post them here? Note that the files I have listed will have a timestamp between the `crash` part and the extension.

**Comment by LC-Zorg** [comment]
I can't upload - the file is too big. But probably you can very easily recreate it yourself by removing GS from the existing scenario in the editor.

Ok, I can't upload dmp, so crash.log itself.
[crash20221025163056.log](https://github.com/OpenTTD/OpenTTD/files/9864038/crash20221025163056.log)

**Comment by Arastais** [comment]
Clarifying here: you can actually change the game script in the scenario editor, and using @James103's workaround, in existing saves as well (which like you said you were aware of). What's important to note is, when switching between scripts or adding one, the bug probably depends on the actual design of the Game Script and how the script handles being initialized mid game, so I think this would show up differently depending on the Game Script used. For example, before #10152, I had to significantly modify my [GlobalObjectiveSystem](https://github.com/Arastais/GlobalObjectiveSystemGS) game script so that it could be initialized in the middle of a game (for example, It erases everything in the goals window to make sure nothing is left over from the last iteration).

As for deleting a script from a save all together, that may be a different story. There's no way to erase something left over by a game script for example like when you switch to another one. So, while this also depends on the script used, it's more based on what the script leaves behind and what it saves in the save file. Thus, this is most likely a much more common pitfall, but I still don't fully see how the game could crash because of this; It's probably worth it to take a look at the particular script being used in your case and how it works.

**Comment by James103** [comment]
@LC-Zorg I managed to reproduce the crash you described.

To reproduce:
1. Start a new game with the "Deliverance" game script.
2. Wait until the storybook pops up.
3. Save the game.
4. Rename the save file extension from `.sav` to `.scn` and move the file to the scenarios folder.
5. Abandon game and open Scenario Editor.
6. Load the scenario created in step 4, or load this scenario: [Unnamed, 1950-01-01.zip](https://github.com/OpenTTD/OpenTTD/files/10033205/Unnamed.1950-01-01.zip)
7. Open Game Script Settings.
8. Click "Select Game Script"
9. Click "(none)"
10. Click "Accept"
11. The game will render a frame showing no active Game Script then crash.

Crash files: [crash20221117155707.zip](https://github.com/OpenTTD/OpenTTD/files/10033218/crash20221117155707.zip)

**Comment by 2TallTyler** [comment]
Cross-posted from #10179 

> Changing file extensions to trick Scenario Editor into thinking a savegame (where the GS has been started) is a scenario (where the GS has not been started) is not supported behavior.
> 
> We might want to handle this more gracefully (maybe exiting to the main menu with an error message instead of just crashing), but I don't think fixing this workaround is the proper solution to https://github.com/OpenTTD/OpenTTD/issues/10108.
